src/queries/invitationsAggregate.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * @name accounts
+ * @method
+ * @memberof Accounts/NoMeteorQueries
+ * @summary Returns accounts optionally filtered by group IDs
+ * @param {Object} context - an object containing the per-request state
+ * @param {String} input - input for query
+ * @param {String} [input.shopIds] - Array of shop IDs to limit the results
+ * @returns {Promise} Mongo cursor
+ */
+export default async function accounts(context, { shopIds }) {
+  const { collections } = context;
+  const { AccountInvites } = collections;
+
+  const pipeline = [];
+
+  if (Array.isArray(shopIds) && shopIds.length > 0) {
+    await Promise.all(shopIds.map(async (shopId) => {
+      await context.validatePermissions("reaction:legacy:invitations", "read", { shopId });
+    }));
+
+    pipeline.push({
+      $match: {
+        shopId: {
+          $in: shopIds
+        }
+      }
+    })
+  } else {
+    await context.validatePermissions("reaction:legacy:invitations", "read");
+  }
+
+  pipeline.push(
```

Comment:
What is the purpose of using an aggregate? If it's only to join the related collections, we usually do that through resolvers rather than aggregate for a few reasons:

1. the queries don't have to happen if the client didn't request those fields
2. every query or mutation that returns an `Invitation` type will get all of the joined data automatically without having to reimplement an aggregate
3. By passing through to `context.query` functions, proper auth checks will be done on joined data

So what I mean is that this would just `return AccountInvites.find()` (possibly with `shopId` filter), and you'd move joins to `Invitation.js` resolver file:

```js
import resolveShopFromShopId from "@reactioncommerce/api-utils/graphql/resolveShopFromShopId.js";
import { encodeInvitationOpaqueId } from "../xforms/id.js";

export default {
  _id: (node) => encodeInvitationOpaqueId(node._id),
  async groups(parent, _, context) {
    const group = await context.queries.groups(context, parent.groupId);
    // If you want to expose this as `groups` to be future proof, that seems fine, but note
    // that it is currently a singular `groupId` field in the db
    return [group];
  },
  invitedBy(parent, _, context) {
    const { invitedByUserId } = parent;
    if (!invitedByUserId) return null;
  
    // Need to create accountByUserId query, exact copy of `userAccount`
    // except query on `userId` field instead of `_id`
    return context.queries.accountByUserId(context, invitedByUserId);
  },
  shop: resolveShopFromShopId
}
```

Reply:
I think that would be at the expense of performance, since aggregates are meant to be way faster than doing multiple queries... But it's true that the resolver approach provides more flexibility. Too bad there isn't a clean way to add the `$lookup` operators dynamically depending on what fields were requested.

That being said, I'll implement your solution. Thanks!

Reply:
We could definitely explore better performance solutions. An aggregate is likely worse for performance if you aren't requesting any subdocuments, but as you said, there could be ways to make it smarter. Bottom line is we debated this stuff long ago and decided that resolvers was the way to do joins. That could change in the future, but we should follow the pattern everywhere for now.

Regarding performance, there is [dataloader](https://github.com/graphql/dataloader) support built into `api-core`, but so far it's only used to optimize a few queries. Lot of potential for future effort on that. (It solves a slightly different problem, but related.)